_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+  spec.executables   = spec.files.grep(%r{^bin/}) { |f| File.basename(f) }
+  spec.test_files    = spec.files.grep(%r{^(test|spec|features)/})
+  spec.require_paths = ["lib"]
+
+  spec.add_development_dependency "bundler", "~> 1.7"
+  spec.add_development_dependency "rake", "~> 10.0"
+end
+{% endhighlight %}
+
+Build the gem
+
+{% highlight sh %}
+gem build prototype_legacy_helper.gemspec
+{% endhighlight %}
+
+Rewind pro and metasploit-framework to commits where pro had ActiveSupport::Dependencies, but framework was still incompatible.  (**NOTE: this state is simulated.  When it happened, framework was updated first before pro was committed, but this state did occur on my dev machine while working on this change.**)
+
+{% highlight sh %}
+cd ~/git/rapid7
+git clone git@rapid7.github.com:rapid7/pro.git
+cd pro
+git checkout  3370204
+rm msf3
+ln -s ~/git/rapid7/metasploit-framework ms3
+cd msf3
+git checkout 8b8da0b
+cd ..
+cd ui
+rvm use ruby-1.9.3@pro
+edit Gemfile
+# Comment out therubyracer.  The locked version can't be built against modern Yosemite
+RAILS_ENV=development bundle install
+{% endhighlight %}
+
+So, everything loads using autoload paths and Metasploit Pro developers no longer have to worry about forgetting to adding requires to prosvc to shared code between `ui` and `engine`, right? Trigger the need to autoload `SocialEngineering` using autoload_paths by removing the constant and then trying to use web_phish module that references `SocialEngineering::Campaign`.  Nope, I get "wrong constant name" when trying to load certain modules for the searching:
+
+![wrong constant name](/images/wrong-constant-name.png)
+
+## Metasploit Module Loader compatibility with ActiveSupport::Dependencies
+
+So, pull up [activesupport-3.2.2/lib/active_support/inflector/methods.rb:229](https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/inflector/methods.rb#L229) to find out how "wrong constant name" is raised:  In Rubymine, open `~/git/rails/rails/activesupport`.  Switch to tag v3.3.2. Shift+Shift `lib/active_support/inflector/methods.rb`. `:229`.
+
+![activesupport-3.2.2/lib/active_support/inflector/methods.rb:229](/images/activesupport-3-2-2-lib-active-support-inflector-methods-line-229.png)
+
+So, since the first line of the backtrace (`/Users/luke.imhoff/.rvm/gems/ruby-1.9.3-p551@pro/gems/activesupport-3.2.2/lib/active_support/inflector/methods.rb:229:in 'const_defined?'`) mentioned `const_defined?`, I assume that `const_defined?` itself raised the exception and `name` is malformed.  I'm assuming `name` is a `String` (and not a symbol or object that responds to `#to_s`) since it comes from `names` and `names` comes from `camel_cased_word`, which calls `split`, which looks like `String#split`.  `#<KLASS:HEX>` is the MRI format for objects that don't have a an overridden `#inspect` method, so I need to figure out how `constantize` is being called on a `Module` that doesn't define `#inspect`.
```

Comment:
So, since the first line of the backtrace (`/Users/luke.imhoff/.rvm/gems/ruby-1.9.3-p551@pro/gems/activesupport-3.2.2/lib/active_support/inflector/methods.rb:229:in 'const_defined?'`) mentioned `const_defined?`, I assume that `const_defined?` itself raised the exception and `name` is malformed.  I'm assuming `name` is a `String` (and not a symbol or object that responds to `#to_s`) since it comes from `names` and `names` comes from `camel_cased_word`, which calls `split`, which looks like `String#split`.  `#<KLASS:HEX>` is the MRI format for objects that don't have **~~a~~** an overridden `#inspect` method, so I need to figure out how `constantize` is being called on a `Module` that doesn't define `#inspect`.